README.md
My Hackerspace
==============

>  **Discontinued** The current development takes place at [spaceapi-community/my-hackerspace](https://github.com/spaceapi-community/my-hackerspace).

- Status of hackerspaces using the [SpaceAPI](https://spaceapi.io/)
- Show information about the space (contact, location, sensors, ...)
- Status widget, multiple widgets supported
- Available on [f-droid](https://f-droid.org/repository/browse/?fdid=ch.fixme.status) and [play store](https://play.google.com/store/apps/details?id=ch.fixme.status)

Master branch: [![Build Status](https://travis-ci.org/fixme-lausanne/MyHackerspace.svg?branch=master)](https://travis-ci.org/fixme-lausanne/MyHackerspace)

HOW TO COMPILE
=============

First, get the sources.

    git clone --recursive https://github.com/fixme-lausanne/MyHackerspace.git
    cd MyHackerspace

Get the 3rd party librairies

    git submodule init
    git submodule update

Android Studio
--------------

With Android Studio, simply open the project directory and you should be set.

Command Line
------------

You can build the project using Gradle.

You'll first need the Android SDK, and install build tools 21.1.1 which is considered obsolete.
You can find this version by ticking obsolete in the Android SDK Manager.

The following examples use the gradle wrapper script which will automatically
download gradle to your local directory. If you want to use your own
system-wide installation instead, simply replace `./gradlew` commands with
`gradle`.

First, copy `local.properties.example` to `local.properties` and adjust the
path to your Android SDK installation.

To build a debug APK:

    ./gradlew assembleDebug

You will find your APK file in the ` app/build/outputs/apk/` directory.

You can also build and directly install the file to your connected smartphone:

    ./gradlew assembleDebug installDebug

To see other tasks that gradle offers, run

    ./gradlew tasks

LOCAL DIRECTORY
===============

For testing purposes you can run a local directory using this technique:

* Create a new Android AVD called for instance "android6"
* Start the AVD from the command line:
    `emulator -avd android6 -shared-net-id 16`
* Make sure a network interface on your host computer is reachable by this IP:
    `sudo ifconfig eth0 10.0.2.3`
* On the host, go to the test directory and run the serv.py script:
    `./serv.py`
* Go in the app preferences and set the OpenSpaceDirectory URL to the following:
    `https://10.0.2.3:8443/directory.json`

TODO
====

- Auto recognize field types in the API (array, obj, string, etc)
- Integrate woozzu library as 3rd party



CHANGELOG.md
# CHANGELOG

Possible tags:

- [info] An information not directly related to code changes
- [feature] A new feature or an improvement
- [bug] A bugfix
- [change] A change that's neither a feature nor a bugfix
- [i18n] Internationalization / translation


# Unreleased


# v2.1.2 (2023-10-02)

- [feature] Show state "last change" timestamp as localized datetime ([#47])
- [bug] Use correct mastodon property ([#49])
- [bug] Fix network error message
- [change] Drop support for Android 5–7, require at least Android 8
- [change] Remove old autolinking-workaround for HTC devices ([#48])

Contributors to this version:

- cyroxx (@cyroxx)
- Danilo Bargen (@dbrgn)

Thanks!

[#47]: https://github.com/spaceapi-community/my-hackerspace/pull/47
[#48]: https://github.com/spaceapi-community/my-hackerspace/pull/48
[#49]: https://github.com/spaceapi-community/my-hackerspace/pull/49


## v2.1.1 (2023-04-27)

- [bug] Fix a bug when parsing v14 endpoints that contain a SpaceFED key
  without a "spacephone" field ([#44])
- [bug] Temporarily use directory from GitHub directly to speed up loading
  ([#45])

Contributors to this version:

- Danilo Bargen (@dbrgn)

Thanks!

[#44]: https://github.com/spaceapi-community/my-hackerspace/pull/44
[#45]: https://github.com/spaceapi-community/my-hackerspace/pull/45


## v2.1.0 (2023-04-16)

- [feature] Add information about the app to settings ([#18])
- [bug] Fix widget crashing on Android 12 ([#24])
- [bug] Export widget config activity to fix crashing OpenLauncher ([#33])
- [change] Reverse latitude and longitude displayed on the screen ([#37])
- [change] Improvements for widget ([#34])
- [change] Upgrade dependencies, change TargetSDK to 33 ([#42])

Contributors to this version:

- Danilo Bargen (@dbrgn)
- Nicco Kunzmann (@niccokunzmann)
- Adrian Pascu (@adipascu)
- Frieder Hannenheim (@FriederHannenheim)

Thanks!

[#18]: https://github.com/spaceapi-community/my-hackerspace/pull/18
[#24]: https://github.com/spaceapi-community/my-hackerspace/pull/24
[#33]: https://github.com/spaceapi-community/my-hackerspace/pull/33
[#34]: https://github.com/spaceapi-community/my-hackerspace/pull/34
[#37]: https://github.com/spaceapi-community/my-hackerspace/pull/37
[#42]: https://github.com/spaceapi-community/my-hackerspace/pull/42
[#44]: https://github.com/spaceapi-community/my-hackerspace/pull/44


## v2.0.2 (2022-08-07)

Unfortunately the app was pulled down from Google Play by Google due to
"developer inactivity". Due to the lack of e-mail forwarding, we never noticed
the warnings, so the app is now gone.

To fix this, we had to change the app ID. Additionally, there will be an F-Droid release.

Changes:

- [change] Update dependencies
- [change] Rename package to `io.spaceapi.community.myhackerspace`
- [info] Add F-Droid metadata


## v2.0.1 (2021-05-14)

- [bug] Fix refresh button ([#5][i5])

[i5]: https://github.com/spaceapi-community/my-hackerspace/pull/5


## v2.0.0 (2021-02-20)

- [info] App was re-released by the SpaceAPI project under a new package name ([#1][i1])
- [info] GitHub is now at https://github.com/spaceapi-community/my-hackerspace/
- [info] The app now requires at least Android 5 (API 21) ([#75][i75])
- [feature] Support for SpaceAPI v14 ([#85][i85])
- [feature] New app launcher icon ([#3][i3])
- [feature] More modern icons in app UI ([#74][i74])
- [bug] Don't save empty data in application state ([#64][i64])
- [change] Update all domains to spaceapi.io ([#65][i65], [#71][i71])
- [change] Switch to Java 8 ([#73][i73])
- [change] Remove MemorizingTrustManager ([#65][i65])
- [change] Upgrade dependencies ([#69][i69])
- [change] Switch to CircleCI ([#69][i69])
- [change] Add support for annotations ([#77][i77])
- [i18n] Improved translations

[i1]: https://github.com/spaceapi-community/my-hackerspace/pull/1
[i3]: https://github.com/spaceapi-community/my-hackerspace/pull/3
[i64]: https://github.com/fixme-lausanne/MyHackerspace/pull/64
[i65]: https://github.com/fixme-lausanne/MyHackerspace/pull/65
[i69]: https://github.com/fixme-lausanne/MyHackerspace/pull/69
[i71]: https://github.com/fixme-lausanne/MyHackerspace/pull/71
[i73]: https://github.com/fixme-lausanne/MyHackerspace/pull/73
[i74]: https://github.com/fixme-lausanne/MyHackerspace/pull/74
[i75]: https://github.com/fixme-lausanne/MyHackerspace/pull/75
[i77]: https://github.com/fixme-lausanne/MyHackerspace/pull/77
[i85]: https://github.com/fixme-lausanne/MyHackerspace/pull/85


## v1.8.3 (2017-01-XX)

- Change links from SpaceAPI.net to SpaceDirectory.org
- Display all webcams


## v1.8.2 (2016-07-02)

- Fix camera and stream url being displayed
- Fix twitter link to the new url format


## v1.8.1 (2016-05-06)

- Uses custom API directory end point (https://spaceapi.fixme.ch/directory.json)
- Allow editing of the API directory end point and the current hackerspace API
- Add Danish translation (thanks Mikkel)


## v1.8 (2016-04-14)

- Supports invalid SSL certificates
- Allow widget to be resized
- Add Dutch translation
- Fix http to https redirection
- General fixes


## v1.7.4.1 (2014-08-26)

- Fix crash when there's no error message


## v1.7.4 (2014-08-07)

- German translation (thanks to Lokke and Phervieux)
- Better hs list with alphabetical index
- Better errors messages
- Caching for http requests (images, hs directory)
- Add status message to the widget (thanks Fpletz)
- Fix bugs: widget updates, ignore ext fields, click from widget


## v1.7.3 (2013-10-25)

- Fix regression with widget custom open/close logo
- Fix order of hackerspaces with different cases


## v1.7.2 (2013-09-09)

- Better layout for sensors
- Support more fields for sensors (machines, names, properties)


## v1.7.1 (2013-09-06)

- Faster http requests (Use DefaultHttpClient instead of HttpURLConnection)


## v1.7 (2013-09-05)

- Full support of SpaceAPI 0.13, drops mixed api definition: hackerspaces must comply to the level they declare!
- Widget transparency preference added (by default transparency is deactivated)



src/ch/fixme/status/ParseGeneric.java
/*
 * Copyright (C) 2013 Aubort Jean-Baptiste (Rorist)
 * Licensed under GNU's GPL 3, see README
 */
package ch.fixme.status;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

public class ParseGeneric {
    protected HashMap<String, Object> mResult = new HashMap<String, Object>();
    protected JSONObject mApi;

    protected static final String API_DIRECTORY = "http://spaceapi.net/directory.json";
    //protected static final String API_DIRECTORY = "https://10.0.2.3:8443/directory.json";
    protected static final String API_NAME = "space";
    protected static final String API_LON = "lon";
    protected static final String API_LOCATION = "location";
    protected static final String API_LAT = "lat";
    protected static final String API_URL = "url";
    protected static final String API_LEVEL = "api";
    protected static final String API_STATE = "state";
    protected static final String API_STATE_MESSAGE = "message";
    protected static final String API_STATUS_TXT = "status";
    protected static final String API_DURATION = "duration";
    protected static final String API_EXT_DURATION = "ext_duration";
    protected static final String API_ADDRESS = "address";
    protected static final String API_CONTACT = "contact";
    protected static final String API_EMAIL = "email";
    protected static final String API_IRC = "irc";
    protected static final String API_JABBER = "jabber";
    protected static final String API_PHONE = "phone";
    protected static final String API_SIP = "sip";
    protected static final String API_TWITTER = "twitter";
    protected static final String API_IDENTICA = "identica";
    protected static final String API_FOURSQUARE = "foursquare";
    protected static final String API_ML = "ml";
    protected static final String API_STREAM = "stream";
    protected static final String API_CAM = "cam";
    protected static final String API_SENSORS = "sensors";
    protected static final String API_EXT = "ext_";
    protected static final String API_RADIATION = "radiation";

    // Sensors
    protected static final String API_VALUE = "value";
    protected static final String API_UNIT = "unit";
    protected static final String API_LOCATION2 = "location";
    protected static final String API_NAME2 = "name";
    protected static final String API_DESCRIPTION = "description";
    protected static final String API_MACHINES = "machines";
    protected static final String API_NAMES = "names";
    protected static final String API_PROPERTIES = "properties";

    // State
    protected static final String API_DEFAULT = "https://fixme.ch/status.json";
    protected static final String API_ICON = "icon";
    protected static final String API_ICON_OPEN = "open";
    protected static final String API_ICON_CLOSED = "closed";
    protected static final String API_LOGO = "logo";
    protected static final String API_STATUS = "open";
    protected static final String API_LASTCHANGE = "lastchange";

    public ParseGeneric(JSONObject jsonObject) {
        mApi = jsonObject;
    }

    public ParseGeneric(String jsonString) throws JSONException {
        mApi = new JSONObject(jsonString);
    }

    public HashMap<String, Object> getData() throws JSONException {
        if ("0.13".equals(mApi.getString(ParseGeneric.API_LEVEL))) {
            mResult = new Parse13(mApi).parse();
        } else if ("0.12".equals(mApi.getString(ParseGeneric.API_LEVEL))) {
            mResult = new Parse12(mApi).parse();
        } else {
            throw new JSONException("API LEVEL NOT SUPPORTED: "
                    + mApi.getString(ParseGeneric.API_LEVEL));
        }
        return mResult;
    }

}
